src/realm/query_expression.hpp

```diff
@@ -971,58 +971,26 @@ class Overloads {
     {
 #ifdef REALM_OLDQUERY_FALLBACK // if not defined, never fallback query_engine; always use query_expression
         // Test if expressions are of type Columns. Other possibilities are Value and Operator.
-        const Columns<R>* left_col = dynamic_cast<const Columns<R>*>(static_cast<Subexpr2<L>*>(this));
+        const Columns<L>* left_col = dynamic_cast<const Columns<L>*>(static_cast<Subexpr2<L>*>(this));
         const Columns<R>* right_col = dynamic_cast<const Columns<R>*>(&right);
 
         // query_engine supports 'T-column <op> <T-column>' for T = {int64_t, float, double}, op = {<, >, ==, !=, <=,
         // >=},
         // but only if both columns are non-nullable, and aren't in linked tables.
-        if (left_col && right_col && std::is_same_v<L, R> && !left_col->is_nullable() && !right_col->is_nullable() &&
-            !left_col->links_exist() && !right_col->links_exist()) {
+        if (left_col && right_col && !left_col->links_exist() && !right_col->links_exist()) {
```

Comment:
Where do we have a test that the 2 subexpressions refer to the same table?

Reply:
We don't seem to have coverage for this here, or even for the single column queries. I will add a check in TwoColumnsNode.

Reply:
I think the check should be here. If one of the expressions contains links, the Compare node will be used.

Reply:
Added a check here as well and a test for it.